File: Logica1.java
import java.util.HashMap;
import java.util.Map;

public class Logica1 {
    public static void main(String[] args) {
        Map<String, String> relaciones = new HashMap<>();
        relaciones.put("Orlando", "Rafael");
        relaciones.put("Rafael", "Felipe");

        boolean resultado = esAbuelo("Orlando", "Felipe", relaciones);
        String abuelo = "Orlando";
        String nieto = "Felipe";

        if (resultado) {
            System.out.println(abuelo + " es abuelo de " + nieto);
        } else {
            System.out.println(abuelo + " no es abuelo de " + nieto);
        }
    }

    public static boolean esPadre(String padre, String hijo, Map<String, String> relaciones) {
        return relaciones.get(padre) != null && relaciones.get(padre).equals(hijo);
    }

    public static boolean esAbuelo(String abuelo, String nieto, Map<String, String> relaciones) {
        for (Map.Entry<String, String> entry : relaciones.entrySet()) {
            if (esPadre(abuelo, entry.getKey(), relaciones) && esPadre(entry.getKey(), nieto, relaciones)) {
                return true;
            }
        }
        return false;
    }
}

File: Logica1.cpp
std::map<std::string, std::string> relaciones = {
    {"Orlando", "Rafael"},
    {"Rafael", "Felipe"}
};

bool es_padre(const std::string& padre, const std::string& hijo) {
    return relaciones.count(padre) && relaciones[padre] == hijo;
}

bool es_abuelo(const std::string& abuelo, const std::string& nieto) {
    for (const auto& [padre, hijo] : relaciones) {
        if (es_padre(abuelo, padre) && es_padre(padre, nieto)) {
            return true;
        }
    }
    return false;
}

int main() {
    std::string abuelo = "Orlando";
    std::string nieto = "Felipe";

    if (es_abuelo(abuelo, nieto)) {
        std::cout << abuelo << " es abuelo de " << nieto << std::endl;
    } else {
        std::cout << abuelo << " no es abuelo de " << nieto << std::endl;
    }

    return 0;
}